test/hotspot/jtreg/compiler/c1/TestCheckIndexC1Intrinsic.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2021, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @author Yi Yang
+ * @bug 8265518
+ * @summary C1 intrinsic support for jdk.internal.util.Preconditions.checkIndex
+ * @requires vm.compiler1.enabled
+ * @library /test/lib
+ * @modules java.base/jdk.internal.util
+ * @run main/othervm -XX:TieredStopAtLevel=1 -Xbatch
+ *                   -XX:CompileCommand=dontinline,*TestCheckIndexC1Intrinsic.check*
+ *                   -XX:CompileCommand=compileonly,*TestCheckIndexC1Intrinsic.check*
+ *                   compiler.c1.TestCheckIndexC1Intrinsic
+ */
+
+package compiler.c1;
+
+import jdk.test.lib.Asserts;
+import jdk.internal.util.Preconditions;
+
+public class TestCheckIndexC1Intrinsic {
+    static int limit = 100;
+
+    private static class MyException extends RuntimeException {
+        public MyException(String msg) {
+            super(msg);
+        }
+    }
+
+    static void check0() {
+        long res = Preconditions.checkIndex(0, 1, null);
+        Asserts.assertEquals((int)res, 0);
+        try {
+            Preconditions.checkIndex(1, 1, null);
+        } catch (IndexOutOfBoundsException e) {
+            // got it!
+        }
```

Comment:
I all places where `Precondition.checkIndex` is expected to throw, an AssertionError should  be generated if it doesn't throw:

```
        try {
            Preconditions.checkIndex(1, 1, null);
            throw new AssertionError("Expected IndexOutOfBoundsException not thrown");
        } catch (IndexOutOfBoundsException e) {
            // got it!
        }
```

Reply:
Yes, it does make sense